external-crates/move/crates/move-trace-format/src/trace_state_machine.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    trace_format::{DataLoad, Effect, Location, Read, TraceEvent, TraceIndex, TraceValue, Write},
+    trace_interface::{Tracer, Writer},
+};
+use core::fmt;
+use move_core_types::annotated_value::MoveValue;
+use std::collections::BTreeMap;
+
+#[derive(Debug, Clone)]
+pub struct TraceState {
+    pub loaded_state: BTreeMap<TraceIndex, MoveValue>,
+    pub operand_stack: Vec<TraceValue>,
+    pub call_stack: BTreeMap<TraceIndex, (BTreeMap<usize, TraceValue>, bool)>,
+}
+
+impl TraceState {
+    pub fn new() -> Self {
+        Self {
+            loaded_state: BTreeMap::new(),
+            operand_stack: vec![],
+            call_stack: BTreeMap::new(),
+        }
+    }
+
+    /// Given a reference "location" return the value it points to.
+    pub fn dereference(&self, location: &Location) -> MoveValue {
+        match location {
+            Location::Local(frame_idx, idx) => {
+                let frame = &self.call_stack[frame_idx];
+                frame.0.get(idx).unwrap().snapshot().clone()
+            }
+            Location::Indexed(loc, _offset) => self.dereference(loc),
+            Location::Global(id) => self.loaded_state[id].clone(),
+        }
+    }
+
+    /// Apply an event to the state machine and update the locals state accordingly.
+    pub fn apply_event(&mut self, event: &TraceEvent) {
+        match event {
+            TraceEvent::OpenFrame { frame, .. } => {
+                let mut locals = BTreeMap::new();
+                for (i, p) in frame.parameters.iter().enumerate() {
+                    // NB: we may fail on the open of a main frame but thats fine since we have
+                    // it in the parameters so we can just pop it and not unwrap it to fail
+                    // silently.
```

Comment:
Is this what this comment is trying to explain?
```suggestion
                    // NB: parameters are passed directly, so we just pop to make sure they aren't also
                    //  left on the operand stack. For the initial call, these pops may (should) fail, but that
                    // is fine as we already have the values in the parameter list.
```